aptos-move/aptos-vm/src/aptos_vm.rs

```diff
@@ -2414,38 +2416,38 @@ impl AptosVM {
         })
     }
 
+    #[allow(clippy::manual_filter)]
     pub fn get_required_deposit(
         &self,
         session: &mut SessionExt,
         resolver: &impl AptosMoveResolver,
-        txn_gas_params: &TransactionGasParameters,
+        _txn_gas_params: &TransactionGasParameters,
         txn_metadata: &TransactionMetadata,
         payload: &TransactionPayload,
-    ) -> Option<u64> {
+    ) -> Result<Option<u64>, VMStatus> {
         match payload {
             TransactionPayload::EntryFunction(entry_func) => {
-                if self.randomness_enabled
-                    && has_randomness_attribute(resolver, session, entry_func).unwrap_or(false)
-                {
-                    let max_execution_gas: Gas = txn_gas_params
-                        .max_execution_gas
-                        .to_unit_round_up_with_params(txn_gas_params);
-                    let max_io_gas: Gas = txn_gas_params
-                        .max_io_gas
-                        .to_unit_round_up_with_params(txn_gas_params);
-                    let cand_0 = txn_metadata.gas_unit_price * (max_execution_gas + max_io_gas)
-                        + txn_gas_params.max_storage_fee;
-                    let cand_1 =
-                        txn_metadata.gas_unit_price * txn_gas_params.maximum_number_of_gas_units;
-                    let required_fee_deposit = min(cand_0, cand_1);
-                    Some(u64::from(required_fee_deposit))
+                if let Some(gas_amount) = self.randomness_api_v0_required_deposit.gas_amount {
+                    if has_randomness_attribute(resolver, session, entry_func).unwrap_or(false) {
+                        if gas_amount != u64::from(txn_metadata.max_gas_amount) {
```

Comment:
It makes more sense to me if we allow user to specify max_gas_amount > required_deposit_amount, but use required_deposit_amount to limit the execution. Otherwise I can imagine lots of the contract will break.